src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/IdentifiersShouldNotContainUnderscores.Fixer.cs

```diff
@@ -1,29 +1,87 @@
+    public sealed class IdentifiersShouldNotContainUnderscoresFixer : CodeFixProvider
     {
-        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray<string>.Empty;
+        public sealed override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(IdentifiersShouldNotContainUnderscoresAnalyzer.RuleId);
 
         public sealed override FixAllProvider GetFixAllProvider()
         {
             // See https://github.com/dotnet/roslyn/blob/master/docs/analyzers/FixAllProvider.md for more information on Fix All Providers
             return WellKnownFixAllProviders.BatchFixer;
         }
 
-        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        public sealed override async Task RegisterCodeFixesAsync(CodeFixContext context)
         {
-            // Fixer not yet implemented.
-            return Task.CompletedTask;
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var model = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
 
+            var node = root.FindNode(context.Span);
+            if (node == null)
+            {
+                return;
+            }
+
+            var symbol = model.GetDeclaredSymbol(node, context.CancellationToken);
+            if (symbol == null)
+            {
+                return;
+            }
+
+            string title = MicrosoftCodeQualityAnalyzersResources.IdentifiersShouldNotContainUnderscoresFixTitle;
+            context.RegisterCodeFix(new MyCodeAction(title,
+                                        async ct => await RenameWithoutUnderscoresAsync(context.Document, symbol, ct).ConfigureAwait(false),
+                                        equivalenceKey: title),
+                                    context.Diagnostics);
+        }
+
+        private static async Task<Document> RenameWithoutUnderscoresAsync(Document document, ISymbol symbol, CancellationToken cancellationToken)
+        {
+            Solution newSolution = await Renamer.RenameSymbolAsync(document.Project.Solution, symbol, GetNewName(symbol.Name), null, cancellationToken).ConfigureAwait(false);
+            return newSolution.GetDocument(document.Id)!;
+        }
+
+        private static string GetNewName(string name)
+        {
+            var builder = new StringBuilder();
+            bool isPreviousUnderscore = false;
+            foreach (char c in name)
+            {
+                if (c == '_')
+                {
+                    isPreviousUnderscore = true;
+                    continue;
+                }
+                builder.Append(isPreviousUnderscore ? char.ToUpper(c, CultureInfo.InvariantCulture) : c);
+                isPreviousUnderscore = false;
+            }
+            return builder.ToString();
+        }
+
+        private class MyCodeAction : DocumentChangeAction
```

Comment:
Please change the base class to be `SolutionChangeAction` as in https://github.com/dotnet/roslyn-analyzers/pull/3352/files#diff-6fee323e5832f875afec52f98a70ba86R65

Reply:
Sounds reasonable.